File: backend/src/voice_clone/schemas/content.py
"""Content generation and library schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from voice_clone.models.content import ContentStatus


class GenerationRequest(BaseModel):
    """Schema for content generation request."""

    voice_clone_id: str
    platform: str
    input_text: str = Field(..., min_length=1)

    # Optional properties
    length: str | None = Field(None, pattern="^(short|medium|long)$")
    tone_override: str | None = None
    target_audience: str | None = None
    cta_style: str | None = None


class GenerationResponse(BaseModel):
    """Schema for content generation response."""

    id: str
    content_text: str
    platform: str
    detection_score: int | None
    detection_breakdown: dict[str, int] | None
    properties_used: dict[str, Any]


class ContentCreate(BaseModel):
    """Schema for creating content directly."""

    voice_clone_id: str | None = None
    platform: str
    content_text: str
    input_text: str
    properties_used: dict[str, Any] = Field(default_factory=dict)
    detection_score: int | None = None
    detection_breakdown: dict[str, Any] | None = None
    tags: list[str] = Field(default_factory=list)


class ContentUpdate(BaseModel):
    """Schema for updating content."""

    content_text: str | None = None
    status: ContentStatus | None = None
    tags: list[str] | None = None


class ContentResponse(BaseModel):
    """Schema for content response."""

    id: str
    user_id: str
    voice_clone_id: str | None
    platform: str
    content_text: str
    input_text: str
    properties_used: dict[str, Any]
    detection_score: int | None
    detection_breakdown: dict[str, Any] | None
    status: ContentStatus
    tags: list[str]
    created_at: datetime
    updated_at: datetime | None

    # Include voice clone name for display
    voice_clone_name: str | None = None

    model_config = {"from_attributes": True}


class ContentListResponse(BaseModel):
    """Schema for paginated content list response."""

    items: list[ContentResponse]
    total: int
    page: int
    per_page: int
    pages: int


class ContentFilterParams(BaseModel):
    """Schema for content filter parameters."""

    voice_clone_id: str | None = None
    platform: str | None = None
    status: ContentStatus | None = None
    search: str | None = None
    sort: str = "created_at"
    sort_order: str = "desc"
    page: int = 1
    per_page: int = 20

File: frontend/src/types/content.ts
/**
 * Content type definitions
 */

export enum ContentStatus {
  DRAFT = "draft",
  READY = "ready",
  PUBLISHED = "published",
  ARCHIVED = "archived",
}

export interface Content {
  id: string
  user_id: string
  voice_clone_id?: string
  platform: string
  content_text: string
  input_text: string
  properties_used: ContentProperties
  detection_score?: number
  detection_breakdown?: DetectionBreakdown
  status: ContentStatus
  tags: string[]
  created_at: string
  updated_at?: string
  voice_clone_name?: string
}

export interface ContentProperties {
  length?: string
  tone_override?: string
  target_audience?: string
  cta_style?: string
  dna_version?: number
}

export interface DetectionBreakdown {
  sentence_variety: number
  vocabulary_diversity: number
  specificity: number
  transition_naturalness: number
  opening_closing: number
  punctuation: number
  personality: number
  structure: number
}

export interface GenerationRequest {
  voice_clone_id: string
  platform: string
  input_text: string
  length?: "short" | "medium" | "long"
  tone_override?: string
  target_audience?: string
  cta_style?: string
}

export interface GenerationResponse {
  id: string
  content_text: string
  platform: string
  detection_score?: number
  detection_breakdown?: DetectionBreakdown
  properties_used: ContentProperties
}

export interface ContentUpdate {
  content_text?: string
  status?: ContentStatus
  tags?: string[]
}

export interface ContentListResponse {
  items: Content[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface ContentFilters {
  voice_clone_id?: string
  platform?: string
  status?: ContentStatus
  search?: string
  sort?: string
  order?: "asc" | "desc"
  page?: number
  per_page?: number
}

export interface PlatformLimit {
  platform: string
  name: string
  char_limit?: number
  supports_threads: boolean
  thread_limit?: number
}

export interface PlatformPreview {
  platform: string
  content: string
  char_count: number
  char_limit?: number
  is_within_limit: boolean
  threads?: string[]
}

export const PLATFORMS = [
  { value: "linkedin", label: "LinkedIn" },
  { value: "twitter", label: "Twitter/X" },
  { value: "facebook", label: "Facebook" },
  { value: "instagram", label: "Instagram" },
  { value: "email", label: "Email" },
  { value: "blog", label: "Blog" },
  { value: "sms", label: "SMS" },
] as const

export type Platform = typeof PLATFORMS[number]["value"]
